cpf.go

```diff
@@ -0,0 +1,112 @@
+package godocs
+
+import (
+	"errors"
+	"fmt"
+	"regexp"
+)
+
+// Constants with CPF informations
+const (
+	CpfLength         = 11
+	CpfRegex          = `^([\d]{3})([\d]{3})([\d]{3})([\d]{2})$`
+	CpfNumberOfDigits = 2
+)
+
+// Cpf base structure
+type Cpf struct {
+	number         string
+	digit          string
+	length         int32
+	numberOfDigits int32
+}
+
+// NewCpf is the constructor of Cpf
+func NewCpf(number string) (Cpf, error) {
+	regex := regexp.MustCompile(`/\D/`)
+	cpf := regex.ReplaceAllString(number, ``)
+
+	document := Cpf{
+		number:         cpf,
+		length:         CpfLength,
+		numberOfDigits: CpfNumberOfDigits,
+	}
+	document.digit = document.ExtractCheckerDigit()
+
+	err := document.Assert()
+	if err != nil {
+		return Cpf{}, err
+	}
+
+	return document, nil
+}
+
+// ToString returns a string with the CPF number unformatted
+func (document Cpf) ToString() string {
+	return document.number
+}
+
+// Format returns a string with the CPF number formatted
+func (document Cpf) Format() string {
+	regex := regexp.MustCompile(CpfRegex)
+	return regex.ReplaceAllString(document.number, "$1.$2.$3-$4")
+}
+
+// CalculateDigit do the calculation of Cpf verificator digit
+func (document Cpf) CalculateDigit() string {
+	calculator := NewDigitCalculator(document.ExtractBaseNumber())
+	calculator.WithMultipliersInterval(CpfNumberOfDigits, CpfLength)
+	calculator.UseComplementaryInsteadOfModule()
+	calculator.ReplaceWhen(0, []int{10, 11})
+	calculator.WithModule(Module11)
+
+	firstDigit := calculator.Calculate()
+	calculator.AddDigit(firstDigit)
+	secondDigit := calculator.Calculate()
+	return fmt.Sprintf("%d%d", firstDigit, secondDigit)
+}
+
+// Assert do the verifications to know if a Cpf is valid and not empty
+func (document Cpf) Assert() error {
+	if document.number == "" {
+		return errors.New("cpf number can not be empty")
+	}
+
+	if !document.isValid() {
+		return errors.New("cpf is not valid")
+	}
+
+	return nil
+}
+
+// ExtractCheckerDigit returns the verificator digit of a CPF
+func (document Cpf) ExtractCheckerDigit() string {
```

Comment:
this method could be generic, what you think create an Interface and keep it agnostic than CPF?

Reply:
I have doubts if is the better approach create interface to keep this kind of function agnostic and more reusable.
As Golang doesn't support aggregation, and this better to move it to an another type/package.

I'm really think if is good to expose this kind of function on API of a document.